### PR TITLE
add soft descriptor support

### PR DESCRIPTION
**Pull request by unieagle**
@sbrossie @pierre Please review, thanks!

### CODE REVIEW

**Comment by unieagle** [comment]
@pierre code updated.

File: lib/cybersource/api.rb [hunk 1]
```diff
@@ -276,6 +279,16 @@ def should_credit?(kb_payment_id, context, options = {})
         (now - transaction.created_at) >= threshold
       end
 
+      def before_gateways(kb_transaction, last_transaction, payment_source, amount_in_cents, currency, options, context = nil)
+        # Provide necessary information for merchant descriptor
+        if options[:merchant_descriptor].present?
+          options[:merchant_descriptor][:is_amex] = (payment_source.brand == 'american_express')
```
Comment (pierre) [inline lib/cybersource/api.rb]: `.brand` may not be defined in case of tokenized card (https://github.com/killbill/killbill-plugin-framework-ruby/blob/master/lib/killbill/helpers/active_merchant/payment_plugin.rb#L495). In such cases, we should:

1. Check if the brand is passed via plugin properties (property `cc_type`, see https://github.com/killbill/killbill-plugin-framework-ruby/blob/master/lib/killbill/helpers/active_merchant/payment_plugin.rb#L177)
2. As a last resort, retrieve the payment method again to see if the brand was stored on disk

File: lib/cybersource/api.rb [hunk 2]
```diff
@@ -276,6 +279,16 @@ def should_credit?(kb_payment_id, context, options = {})
...
+          options[:merchant_descriptor][:transaction_type] = kb_transaction.transaction_type
```
Comment (pierre) [inline lib/cybersource/api.rb]: It looks like it's set twice (see `add_merchant_descriptor` below)?

File: lib/cybersource/api.rb [hunk 3]
```diff
@@ -336,6 +349,15 @@ def add_required_options(kb_account_id, properties, options, context)
         ::Killbill::Plugin::ActiveMerchant::Utils.normalize_property(properties, 'ignore_cvv')
       end
 
+      def add_merchant_descriptor(transaction_type, properties, options)
+        merchant_descriptor = find_value_from_properties(properties, 'merchant_descriptor')
+        merchant_descriptor_hash = JSON.parse(merchant_descriptor) rescue nil
+        unless merchant_descriptor_hash.nil?
+          merchant_descriptor_hash[:transaction_type] = transaction_type
```
Comment (pierre) [inline lib/cybersource/api.rb]: See remark above. Also, this is going to be a symbol.

File: lib/cybersource/ext/active_merchant/active_merchant.rb [hunk 4]
```diff
@@ -129,6 +129,30 @@ def add_merchant_data(xml, options)
...
+
+      def add_invoice_header(xml, options)
+        merchant_descriptor = options[:merchant_descriptor]
+        if merchant_descriptor.present? #merchant_descriptor!=nil && !merchant_descriptor.empty?
+          is_amex = merchant_descriptor[:is_amex]
+          type    = merchant_descriptor[:transaction_type]
+          name    = merchant_descriptor[:name]
+          contact = merchant_descriptor[:contact]
+          if is_amex
+            if type != :AUTHORIZE # Amex only support capture and refund
```
Comment (pierre) [inline lib/cybersource/ext/active_merchant/active_merchant.rb]: Nit: support**s**

File: spec/cybersource/base_plugin_spec.rb [hunk 5]
```diff
@@ -145,6 +145,32 @@
...
+      end
+      purchase_with_token(:PROCESSED, [], expected_successful_params)
+    end
+
+    it 'can provide merchant descriptor for visa' do
+      ::ActiveMerchant::Billing::CyberSourceGateway.any_instance.stub(:ssl_post) do |host, request_body|
+        request_body.should match('<invoiceHeader>\n        <merchantDescriptor>Ray Qiu               </merchantDescriptor>\n        <merchantDescriptorContact>650-888-3161</merchantDescriptorContact>\n      </invoiceHeader>')
+        successful_purchase_response
+      end
+      purchase_with_card(:PROCESSED, [build_property('merchant_descriptor', {"name"=>"Ray Qiu", "contact"=>"6508883161"}.with_indifferent_access)], expected_successful_params)
```
Comment (pierre) [inline spec/cybersource/base_plugin_spec.rb]: Could you add tests with merchant descriptor against a token and Apple Pay?

**Comment by pierre** [comment]
General question: were you able to run the regression tests against the sandbox (`rake test:remote:spec`)?

File: lib/cybersource/api.rb (hunk 1 above)
Comment (unieagle) [inline lib/cybersource/api.rb]: ok, let me improve it.

File: lib/cybersource/api.rb (hunk 2 above)
Comment (unieagle) [inline lib/cybersource/api.rb]: removed those code in `add_merchant_descriptor`, only set them here.

Comment (unieagle) [inline lib/cybersource/api.rb]: good catch! will update it.

File: lib/cybersource/api.rb (hunk 3 above)
Comment (unieagle) [inline lib/cybersource/api.rb]: good catch!

File: lib/cybersource/ext/active_merchant/active_merchant.rb (hunk 4 above)
Comment (unieagle) [inline lib/cybersource/ext/active_merchant/active_merchant.rb]: good catch!

File: spec/cybersource/base_plugin_spec.rb (hunk 5 above)
Comment (unieagle) [inline spec/cybersource/base_plugin_spec.rb]: sure, will do.

**Comment by unieagle** [comment]
@pierre can not test via `rake test:remote:spec`, got:
```
Unsuccessful add_payment_method: {"exception_message":"wsse:InvalidSecurity: \nSecurity Data : illegal null input\n","payment_plugin_status":"CANCELED"}
```

**Comment by unieagle** [comment]
@pierre will improve this PR again once find a path to retrieve the card type in case there is only token available. May discuss with you next week.

**Comment by pierre** [comment]
> @pierre can not test via `rake test:remote:spec`

It looks like you didn't configure your credentials (see the [cybersource.yml](https://github.com/killbill/killbill-cybersource-plugin/blob/master/cybersource.yml)).

> find a path to retrieve the card type in case there is only token available

In case there is only a token available, your options are (as mentioned in my comment above):

1. Look-up the plugin properties (does the API user triggering the payment know the card type and even though he is relying on the token, he passes the card type nonetheless as a hint)
2. Look-up the `cybersource_payment_methods` entry (your original patch): when the card was tokenized, the card type was most likely known

There is a third option which is to retrieve the card type from CyberSource vault. This has some PCI implications though (and you probably don't need it).

**Comment by unieagle** [comment]
@pierre add some integration tests, now all the tests are green;
made the change that only send soft descriptor when all the information is available, I think currently this is enough, since GDT will send full credit card information to KB, no cybersource token involved.
The currently use case is that credit card information will stored in the `billing_record_responses` table, is this method good for this case? Thanks!

File: spec/cybersource/base_plugin_spec.rb [hunk 6]
```diff
@@ -196,6 +196,83 @@
...
+      context 'amex' do
+        let(:card_type){ :amex }
+        let(:invoice_match_status){ :except_authorize }
+        it_behaves_like 'full payment'
+      end
+    end
+  end
+
+  context 'Invoice Header' do
+    let(:add_payment_prpperties){ :add_card_property }
```
Comment (pierre) [inline spec/cybersource/base_plugin_spec.rb]: Typo: `add_payment_properties`.

File: spec/cybersource/base_plugin_spec.rb [hunk 7]
```diff
@@ -215,43 +292,97 @@ def with_transaction(kb_payment_id, transaction_type, created_at, context)
...
+    else
+      properties << build_property('cc_number', '4111111111111111')
+      properties << build_property('brand', 'visa')
+    end
     properties << build_property('email', 'foo@bar.com')
-    properties << build_property('cc_number', '4111111111111111')
-    properties << build_property('brand', 'visa')
     properties << build_property('eci', '05')
-    properties << build_property('payment_cryptogram', '111111111100cryptogram')
+    properties << build_property('payment_cryptogram', Base64.encode64('111111111100cryptogram'))
```
Comment (pierre) [inline spec/cybersource/base_plugin_spec.rb]: Why this change?

File: spec/cybersource/remote/integration_spec.rb [hunk 8]
```diff
@@ -545,6 +545,42 @@
...
   end
 
+  shared_examples 'success_auth_capture_and_refund' do
+    it 'should be able to auth, capture and refund with descriptors' do
+      @pm = create_payment_method(::Killbill::Cybersource::CybersourcePaymentMethod, nil, @call_context.tenant_id, @properties)
+
+      payment_response = @plugin.authorize_payment(@pm.kb_account_id, payment_id, SecureRandom.uuid, @pm.kb_payment_method_id, @amount, @currency, @properties, @call_context)
+      check_response(payment_response, @amount, :AUTHORIZE, :PROCESSED, 'Successful transaction', '100')
+
+      # Try a captures
```
Comment (pierre) [inline spec/cybersource/remote/integration_spec.rb]: Typo: `capture`.

File: lib/cybersource/api.rb [hunk 9]
```diff
@@ -276,6 +279,20 @@ def should_credit?(kb_payment_id, context, options = {})
         (now - transaction.created_at) >= threshold
       end
 
+      def before_gateways(kb_transaction, last_transaction, payment_source, amount_in_cents, currency, options, context = nil)
+        # Provide necessary information for merchant descriptor
+        if options[:merchant_descriptor].present? && options[:merchant_descriptor]['is_amex'].nil?
```
Comment (pierre) [inline lib/cybersource/api.rb]: My [remark](https://github.com/killbill/killbill-cybersource-plugin/pull/7#discussion_r70713412) still holds:

* Check the plugin property `cc_type` instead of making the user specify another property inside the descriptor
* Check if `respond_to?(:brand)` instead below, for performance reasons (the vast majority of calls would throw an unnecessary exception otherwise)

File: lib/cybersource/api.rb [hunk 10]
```diff
@@ -336,6 +353,15 @@ def add_required_options(kb_account_id, properties, options, context)
         ::Killbill::Plugin::ActiveMerchant::Utils.normalize_property(properties, 'ignore_cvv')
       end
 
+      def add_merchant_descriptor(transaction_type, properties, options)
+        merchant_descriptor = find_value_from_properties(properties, 'merchant_descriptor')
+        merchant_descriptor_hash = JSON.parse(merchant_descriptor) rescue nil
```
Comment (pierre) [inline lib/cybersource/api.rb]: For performance reasons, check if `merchant_descriptor` is `blank` before trying to parse it (otherwise the majority of calls will throw an unnecessary exception).

File: spec/cybersource/base_plugin_spec.rb [hunk 11]
```diff
@@ -196,6 +196,83 @@
...
+        it_behaves_like 'full payment'
+      end
+    end
+  end
+
+  context 'Invoice Header' do
+    let(:add_payment_prpperties){ :add_card_property }
+
+    context 'payments with card' do
+      it_behaves_like 'invoice header example'
```
Comment (pierre) [inline spec/cybersource/base_plugin_spec.rb]: Shouldn't the `let` be in the context, similar to the scenario below?

**Comment by pierre** [comment]
> add some integration tests, now all the tests are green

👍

File: spec/cybersource/base_plugin_spec.rb (hunk 6 above)
Comment (unieagle) [inline spec/cybersource/base_plugin_spec.rb]: Good catch! You review so carefully 👍

File: spec/cybersource/base_plugin_spec.rb (hunk 7 above)
Comment (unieagle) [inline spec/cybersource/base_plugin_spec.rb]: According to https://github.com/killbill/killbill-cybersource-plugin/blob/master/lib/cybersource/ext/active_merchant/active_merchant.rb#L63
Found there are some difference between amex and other cards, so for amex, there is a more decode step.
Code updated in the next commit.

File: spec/cybersource/remote/integration_spec.rb (hunk 8 above)
Comment (unieagle) [inline spec/cybersource/remote/integration_spec.rb]: 👍 code updated in the next commit.

File: lib/cybersource/api.rb (hunk 10 above)
Comment (unieagle) [inline lib/cybersource/api.rb]: good point, checked in the next commit, but I kept the rescue logic, because `JSON.parse` is vulnerable to the string format.

File: spec/cybersource/base_plugin_spec.rb (hunk 11 above)
Comment (unieagle) [inline spec/cybersource/base_plugin_spec.rb]: the `let` in line 264 created a default setting of `:add_payment_properties` in all the context within line 263. but since there are only 2 context within context line 263, it looks better if put the `let` into context at line 266.
updated this in the next commit.

File: lib/cybersource/api.rb (hunk 9 above)
Comment (unieagle) [inline lib/cybersource/api.rb]: Hi @pierre ,
This is the most difficult part of this PR, for our use case, the credit card information will only passed through the authorize call (actually the combo call), and no card information will be in the capture, refund and void call. In this case, I think the cc_type will be empty in the transaction requests other than authorize. If this is true, then I think check `cc_type` is not worthy.
For `respond_to?(:brand)`, will update the code in the next commit.

Comment (pierre) [inline lib/cybersource/api.rb]: > This is the most difficult part of this PR, for our use case, the credit card information will only passed through the authorize call (actually the combo call), and no card information will be in the capture, refund and void call.

Understood, but I don't think the user of the API should have to bother filling-in these details. Let's chat offline about it.

File: lib/cybersource/api.rb [hunk 12]
```diff
@@ -336,6 +352,22 @@ def add_required_options(kb_account_id, properties, options, context)
...
+        return unless merchant_descriptor.present?
+        
+        merchant_descriptor_hash = JSON.parse(merchant_descriptor) rescue nil
+        return unless merchant_descriptor_hash.present? && merchant_descriptor_hash.is_a?(Hash)
+
+        merchant_descriptor_hash = merchant_descriptor_hash.with_indifferent_access
+
+        merchant_descriptor_hash[:transaction_type] = transaction_type
+        if merchant_descriptor_hash[:card_type].nil?
+          merchant_descriptor_hash[:card_type] = find_value_from_properties(properties, 'cc_type')
```
Comment (unieagle) [inline lib/cybersource/api.rb]: Hi @pierre , is this the correct way of retrieve `cc_type` from properties?
This is the first place to get cc_type from input, and in the `before_gateways` will be the last resort to populate this value.

**Comment by unieagle** [comment]
@pierre code updated.

**Comment by pierre** [comment]
Thanks! 👍
